enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,398 @@
+* Minimize any differences to the primary OpenShift platforms
+* Avoid any decisions that would prevent future implementation and support for upgrade/downgrade paths between two-node and traditional architectures 
+* Provide an OpenShift cluster experience that is identical to that of a 3-node hyperconverged cluster, but with 2 nodes
+
+### Non-Goals
+
+* Workload resilience - see related [Pre-DRAFT enhancement](https://docs.google.com/document/d/1TDU_4I4LP6Z9_HugeC-kaQ297YvqVJQhBs06lRIC9m8/edit)
+* Resilient storage - see future enhancement
+* Support for platforms other than bare metal including automated ci testing
+* Support for other topologies (eg. hypershift)
+* Adding worker nodes
+* Creation RHEL-HA events and metrics for consumption by the OpenShift monitoring stack (Deferred to post-MVP)
+
+## Proposal
+
+Use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to delivered supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
+etcd will run as as a voting member on both nodes.
+We will take advantage of RHEL-HA's native support for systemd and re-use the standard cri-o and kublet units, as well as create a new Open Cluster Framework (OCF) script for etcd.
+The existing startup order of cri-o, then kubelet, then etcd will be preserved.
+The `etcdctl`, `etcd-metrics`, and `etcd-readyz` containers will remain part of the static pod, the contents of which remains under the exclusive control of the Cluster Etcd Operator (CEO).
+
+Use RedFish compatible Baseboard Management Controllers (BMCs) as our primary mechanism to power off (fence) an unreachable peer and ensure that it can do no harm while the remaining node continues.
+
+Upon a peer failure, the RHEL-HA components on the surivor will fence the peer and use the OCF script to restart etcd as a new cluster-of-one.
+
+Upon a network failure, the RHEL-HA components ensure that exactly one node will survive, fence it's peer, and use the OCF script to restart etcd as a new cluster-of-one.
+
+In both cases, the control-plane's dependance on etcd will cause it to respond with errors until etcd has been restarted.
+
+Upon rebooting, the RHEL-HA components ensure that a node remains inert (not running cri-o, kubelet, or etcd) until it sees it's peer.
+If the failed peer is likely to remain offline for an extended period of time, admin confirmation is required on the remaining node to allow it to start OpenShift.
+The functionality exists within RHEL-HA, but a wrapper will be provided to take care of the details.
+
+When starting etcd, the OCF script will use etcd's cluster ID and version counter to determine whether the existing data directory can be reused, or must be erased before joining an active peer.
+
+### Workflow Description
+
+#### Cluster Creation
+
+User creation of a two node control-plane will be possible via the Assisted Installer.
```

Comment:
Is this maybe a typo - since later it's mentioned the usage of ABI?

Reply:
I don't think it's a typo, but I also don't think it's necessary to call out Assisted Installer here at all (in any of its forms). This enhancement should focus specifically on what is required to create a 2-node cluster in the absence of any additional automation.

Assuming this is done I don't see any problem so far that would prevent us from allowing 2-node installs from ABI, Assisted SaaS, or MCE

Reply:
Not sure I entirely agree. With an external bootstrap the number of etcd replicas during bootstrapping goes from 1->2->3->2, whereas with in-place bootstrap it goes from 1->2->1->2.
So the in-place bootstrap used by assisted requires solving a fundamentally different problem. (IMHO we should focus on this problem and ignore the external bootstrap case. I've spoken about this with beekhof and I think he still disagrees with me :slightly_smiling_face:)
I agree the particular flavour of Assisted doesn't matter as much. The main thing that is potentially different is the API that accepts the BMC details (including credentials), as discussed in https://github.com/openshift/enhancements/pull/1675#discussion_r1802341759

Reply:
I think I'm missing how this is different from the compact case where we have the same difference in behavior.

Without assisted you get 1->2->3->4->3, and with assisted you get 1->2->3->2->3, right? Why would we treat two node differently?

Alternatively I've seen it suggested to use some of what SNO (without assisted) does to achieve a two node cluster, without assisted, with only two nodes, but in either case adding assisted is then a detail that can be handled in the same way it is for existing install methods, right?

Really all I'm saying is that none of the existing install footprints exist knowing that they will be installed with assisted, and I don't think that's an assumption we should add here without an exceedingly good reason. Maybe I'm not fully understanding your suggestions though or maybe we do have that reason, but that hasn't been made clear to me yet.

Reply:
In the  2->3->2->3 part you never have less than 1+n/2 nodes so you can't lose quorum.

In the _general_ case, going from 2->1->2 is unsafe and requires manual intervention, as noted elsewhere in this doc. During bootstrap the second node joining is not the same one that went away, so I don't think it's _that_ serious, but I presume it will require special handling in the pacemaker resource agent for etcd.

I agree with the rest of your comments. (AFAIK though the only ways to installer SNO are via assisted or a UPI-like method using `create single-node-ignition-config` instead of `create ignition-configs`, and I'm not actually sure that we _support_ the UPI-like way.)

Reply:
So I've been trying out the install paths and can confirm a few things:
1. For SNO, we don't enforce protection of etcd scaling during or after install.
2. For compact deployed with an in-memory bootstrap like ABI, the we set a special annotation for etcd to enforce the 3 etcd replica minimum *after* the bootstrapping completes. As noted, you're bouncing from 3 to 2 to 3 so you're never in danger of losing equorum.
3. For 2NO we will probably want to rework our existing strategies to include one that enforces 2 etcd for bootstrap and post deployment (similiar to the default HA strategy but with one less node) and one that allows for single-etcd scaling states during bootstrapping, but enforces the 2 etcd minimum once installed (for ABI).

For development and testing purposes, we can use the wild-west SNO scaling policy for now.